### ISSUE TITLE
Relicense under dual MIT/Apache-2.0

### ISSUE THREAD

**Issue by emberian** [carrotsrc/rsnl#3]
# Why?

The MIT license requires reproducing countless copies of the same copyright
header with different names in the copyright field, for every MIT library in
use. The Apache license does not have this drawback, and has protections from
patent trolls and an explicit contribution licensing clause. However, the
Apache license is incompatible with GPLv2. This is why Rust is dual-licensed as
MIT/Apache (the "primary" license being Apache, MIT only for GPLv2 compat), and
doing so would be wise for this project. This also makes this crate suitable
for inclusion in the Rust standard distribution and other project using dual
MIT/Apache.
# How?

To do this, get explicit approval from each contributor of copyrightable work
(as not all contributions qualify for copyright) and then add the following to
your README:

```
## License

Licensed under either of
 * Apache License, Version 2.0 ([LICENSE-APACHE](LICENSE-APACHE) or http://www.apache.org/licenses/LICENSE-2.0)
 * MIT license ([LICENSE-MIT](LICENSE-MIT) or http://opensource.org/licenses/MIT)
at your option.

### Contribution

Unless you explicitly state otherwise, any contribution intentionally submitted
for inclusion in the work by you shall be dual licensed as above, without any
additional terms or conditions.
```

and in your license headers, use the following boilerplate (based on that used in Rust):

```
// Copyright (c) 2015 t developers
// Licensed under the Apache License, Version 2.0
// <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT
// license <LICENSE-MIT or http://opensource.org/licenses/MIT>,
// at your option. All files in the project carrying such
// notice may not be copied, modified, or distributed except
// according to those terms.
```

And don't forget to update the `license` metadata in your `Cargo.toml`!
# Contributor checkoff
- [x] @carrotsrc
- [x] @RAOF

**Comment by RAOF** [comment]
I'm fine with this, but it's worth noting that this links to the LGPL'd libnl library. Code using this can't _quite_ use it as if it were otherwise unencumbered Apache/MIT.

**Comment by emberian** [comment]
@RAOF for code linking against LGPL crates, I've been recommending the change still, so at least the wrapper code can benefit.

**Comment by RAOF** [comment]
Sure.

And the people who really care about licensing will probably find the LGPL dependency anyway. It might be nice to mention it somewhere, though :smile:

**Comment by emberian** [comment]
Definitely! One thing I thought about was breaking out a -sys crate that had the same license as the underlying library, but since Rust crates are typically statically linked, that can cause gnashing of teeth when the underlying license is LGPL. I don't have a good solution to help automated tools there -- if only SPDX were more widely adopted!

**Comment by carrotsrc** [comment]
Sorry for delay, I'm a bit taken up at the moment but sounds good mate, I have no issues with relicensing

**Comment by emberian** [comment]
My scripts have reported that they cannot open a relicensing pull request
automatically for this repository. Sorry about that! There's a simple script
here that you could use:

https://gist.github.com/kstep/f9f99cca7ae9cc3d2e16

Though be sure to change "The Rust Project Developers" to the appropriate text
in LICENSE-MIT.

Alternatively, I can create the PR manually. Either works for me, just let me
know!
